### PR TITLE
[govee] Fix invalid status response handling

### PR DESCRIPTION
**Pull request by lsiepel**
- Fixes https://github.com/openhab/openhab-addons/issues/17041
- Improve logging a bit
- Add some basic tests

### CODE REVIEW

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java [hunk 1]
```diff
@@ -224,7 +225,9 @@ public void run() {
                                     }
                                 }
                             } catch (JsonParseException e) {
-                                // this probably was a status message
+                                logger.debug(
+                                        "JsonParseException when trying to parse the response, probably a status message");
+                                logger.trace("Response >> {}", response);
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:230]: This is already logged in line 210.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java [hunk 2]
```diff
@@ -224,7 +225,9 @@ public void run() {
                                     }
                                 }
                             } catch (JsonParseException e) {
-                                // this probably was a status message
+                                logger.debug(
+                                        "JsonParseException when trying to parse the response, probably a status message");
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:229]: It might be useful to log the exception details as well:
```suggestion
                                        "JsonParseException when trying to parse the response, probably a status message, e");
```

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java [hunk 3]
```diff
@@ -318,8 +322,13 @@ public void updateDeviceState(@Nullable StatusResponse message) {
         if (newColorTempInKelvin != lastColorTempInKelvin) {
             logger.trace("Color-Temperature Status: old: {} K {}% vs new: {} K", lastColorTempInKelvin,
                     newColorTempInPercent, newColorTempInKelvin);
-            updateState(CHANNEL_COLOR_TEMPERATURE_ABS, new QuantityType<>(lastColorTempInKelvin, Units.KELVIN));
-            updateState(CHANNEL_COLOR_TEMPERATURE, new PercentType(newColorTempInPercent));
+            try {
+                updateState(CHANNEL_COLOR_TEMPERATURE_ABS, new QuantityType<>(lastColorTempInKelvin, Units.KELVIN));
+                updateState(CHANNEL_COLOR_TEMPERATURE, new PercentType(newColorTempInPercent));
+            } catch (IllegalArgumentException e) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:328]: Perhaps each state update should be wrapped separately into try/catch, so that failure to update color temperature in Kelvin doesn't prevent updating color temperature in percent?

I'm also wondering if the handler should verify the values itself, i.e. preventing the issue, rather than deliberately providing invalid values just to have `IllegalArgumentException` thrown. The binding defines the valid range as 2000-9000:
https://github.com/openhab/openhab-addons/blob/35add30e0c9e69d4d246725d207ee45ced80c59b/bundles/org.openhab.binding.govee/src/main/resources/OH-INF/thing/thing-types.xml#L28

so this could be easily validated.

Also, if the received color temperature in Kelvin is something like 9096 (from the issue), perhaps the defined range here is wrong?

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for the fix. I have added a few questions.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java (hunk 2 above)
Comment (maniac103) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:229]: You mean

```
"JsonParseException when trying to parse the response, probably a status message", e); 
```

;-)

Comment (jlaur) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:229]: Yes. 😄

File: bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        final ThingHandlerCallback callback = mock(ThingHandlerCallback.class);
+        final GoveeHandler handler = createAndInitHandler(callback, thing);
+
+        // inject StatusResponseMessage
+        StatusResponse statusMessage = GSON.fromJson(invalidValueJsonString, StatusResponse.class);
+        handler.updateDeviceState(statusMessage);
+
+        try {
+            verify(callback).statusUpdated(eq(thing), argThat(arg -> arg.getStatus().equals(ThingStatus.UNKNOWN)));
+            // verify(callback).statusUpdated(eq(thing), argThat(arg -> arg.getStatus().equals(ThingStatus.ONLINE)));
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:146]: Fixed

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java [hunk 5]
```diff
@@ -110,9 +112,11 @@ public class GoveeHandler extends BaseThingHandler {
         }
     };
 
-    public GoveeHandler(Thing thing, CommunicationManager communicationManager) {
+    public GoveeHandler(Thing thing, CommunicationManager communicationManager,
+            @Nullable ScheduledExecutorService executorService) {
         super(thing);
         this.communicationManager = communicationManager;
+        this.executorService = executorService == null ? this.scheduler : executorService;
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:119]: Fixed. It's intention is to make the class testable. I forgot to change it at the `scheduleWithFixedDelay` call. Now i did.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java (hunk 1 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:230]: Removed.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java (hunk 2 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:229]: Fixed.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 3 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:328]: Changed it so that each update has its own try/catch block.

Previously i considered to pre-check values, but decided not to. Now iyou mention it, i changed my mind and removed the try block for local validation.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:119]: In that case you might want to create a subclass for test purposes. `scheduler` is protected, so you should be able to do that. This way you can keep the real handler "clean" and provide what is needed specifically for test purposes.

[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:117]: It's just a constructor that allows improved testing. `this.scheduler` is also marked final, it can't be set from a subclass.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java [hunk 6]
```diff
@@ -318,12 +320,16 @@ public void updateDeviceState(@Nullable StatusResponse message) {
         if (newColorTempInKelvin != lastColorTempInKelvin) {
             logger.trace("Color-Temperature Status: old: {} K {}% vs new: {} K", lastColorTempInKelvin,
                     newColorTempInPercent, newColorTempInKelvin);
-            updateState(CHANNEL_COLOR_TEMPERATURE_ABS, new QuantityType<>(lastColorTempInKelvin, Units.KELVIN));
-            updateState(CHANNEL_COLOR_TEMPERATURE, new PercentType(newColorTempInPercent));
-        }
+            if (lastColorTempInKelvin >= 2000 && lastColorTempInKelvin <= 9000) {
```
Comment (maniac103) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:323]: `newColorTenpInKelvin`

Besides that, wouldn't clamping the new color temperature/brightness be the better idea compared just preventing the state update? Say the update happened from 2000 to 9001 ... in that case I'd say a state of 9000 is more correct than a state of 2000.

Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:323]: when coding this fix i decided to do no update when the value is invalid. Auto-correcting values is a slippery slope and can become inimitable to users.
This PR is just about reducing the impact on an external bug from Govee.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:117]: You are right, I missed `final`. You could reduce the changes in the production handler to something like this:
```java
protected ScheduledExecutorService executorService = scheduler;
```

Then:
```java
package org.openhab.binding.govee.internal;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.doAnswer;

import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import org.eclipse.jdt.annotation.NonNullByDefault;
import org.mockito.Mockito;
import org.mockito.invocation.InvocationOnMock;
import org.openhab.core.thing.Thing;

@NonNullByDefault
public class GoveeHandlerPartialMock extends GoveeHandler {

    public GoveeHandlerPartialMock(Thing thing) {
        super(thing, new CommunicationManager());

        executorService = Mockito.mock(ScheduledExecutorService.class);
        doAnswer((InvocationOnMock invocation) -> {
            ((Runnable) invocation.getArguments()[0]).run();
            return null;
        }).when(executorService).scheduleWithFixedDelay(any(Runnable.class), anyLong(), anyLong(), any(TimeUnit.class));
    }
}
```

I'm not sure if `CommunicationManager` should also be mocked, is this expected?
```
[INFO] Running org.openhab.binding.govee.internal.GoveeSerializeGoveeHandlerTest
[main] WARN  o.o.c.thing.binding.BaseThingHandler - Handler GoveeHandlerPartialMock tried updating the thing status although the handler was already disposed.
[GoveeStatusReceiver] WARN  o.o.b.g.i.CommunicationManager - exception when receiving status packet
java.net.SocketException: Socket closed
        at java.base/sun.nio.ch.DatagramSocketAdaptor.receive(DatagramSocketAdaptor.java:252)
        at java.base/java.net.DatagramSocket.receive(DatagramSocket.java:700)
        at org.openhab.binding.govee.internal.CommunicationManager$StatusReceiver.run(CommunicationManager.java:203)
Caused by: java.nio.channels.AsynchronousCloseException: null
        at java.base/sun.nio.ch.DatagramChannelImpl.endRead(DatagramChannelImpl.java:523)
        at java.base/sun.nio.ch.DatagramChannelImpl.trustedBlockingReceive(DatagramChannelImpl.java:677)
        at java.base/sun.nio.ch.DatagramChannelImpl.blockingReceive(DatagramChannelImpl.java:635)
        at java.base/sun.nio.ch.DatagramSocketAdaptor.receive(DatagramSocketAdaptor.java:240)
        ... 2 common frames omitted
```

File: bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        }).when(executorStub).scheduleWithFixedDelay(any(Runnable.class), anyLong(), anyLong(), any(TimeUnit.class));
+
+        final GoveeHandler handler = spy(new GoveeHandler(thing, new CommunicationManager(), executorStub));
+
+        handler.setCallback(callback);
+        handler.initialize();
+        return handler;
+    }
+
+    private static State getState(final int input) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:108]: This is unused, can you check the SAT report?

Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:108]: Was intentionally, but ok. I'll also fix the other warnings not related to this PR.

Comment (jlaur) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:108]: Maybe that would be better to do in a separate PR in case we should backport this one, WDYT?

Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:108]: You are right, i'll leave those out to reduce the risk.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java [hunk 8]
```diff
@@ -318,12 +317,16 @@ public void updateDeviceState(@Nullable StatusResponse message) {
...
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: Minor, but perhaps these values could be extracted to constants in `GoveeBindingConstants` for better visibility, since they need to be kept in sync with the same values provided in the thing type declaration.

And then I checked: They are already there: `COLOR_TEMPERATURE_MIN_VALUE` and `COLOR_TEMPERATURE_MAX_VALUE`.

Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: Perhaps the state should otherwise be updated to `UNDEF` to invalidate the previous state after receiving an invalid value?

**Review by jlaur — CHANGES_REQUESTED** [review]
Sorry, hopefully last iteration, found additional minor things that could be improved.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 6 above)
Comment (maniac103) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: So in my example you'd rather accept/expect a reading of 2000 instead of 9000? Or in the case of ramping up brightness, which is then reported as 101, rather keep the old value of say 1 instead of reporting 100?
Can't say I agree there, in Ecovacs binding at least I did the clamping to the valid range.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 8 above)
Comment (maniac103) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: That's a lot better than keeping the old state in my book. In this specific case, which looks like rounding or off-by-one issues to me, I'd prefer clamping though (as mentioned), but setting to UNDEF would be OK to me as well.

Comment (maniac103) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: And, just to be extra clear: this check still uses the wrong variable (old state, not new state).

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java]: Created a mock for it and don't see the exception anymore.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 8 above)
Comment (jlaur) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: I personally agree that rounding to the valid range would probably be preferable (assuming that a value like 9001 is not completely random, but rather meant at close to 9000), but it was rejected previously. As next choice `UNDEF` would be fine too, and just completely ignoring and keeping old state is probably the least desirable.

Comment (jlaur) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: @lsiepel - that probably means that the current test is not effective?

Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: Ok. Thought about it and changed the logic.

I really don't like interpreting/clamping invalid values. But ok. This binding already corrected the low side of the range. So i adjusted it and now it also corrects it at the upper side.

Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:320]: See other comment, no longer relevant.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java [hunk 9]
```diff
@@ -101,7 +102,7 @@ public void sendRequest(GoveeHandler handler, GenericGoveeRequest request) throw
         final byte[] data = message.getBytes();
         final InetAddress address = InetAddress.getByName(hostname);
         DatagramPacket packet = new DatagramPacket(data, data.length, address, REQUEST_PORT);
-        // logger.debug("Sending {} to {}", message, hostname);
+        logger.debug("Sending {} to {}", message, hostname);
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:105]:
```suggestion
        logger.trace("Sending {} to {}", message, hostname);
```

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java [hunk 10]
```diff
@@ -100,9 +101,7 @@ public class GoveeHandler extends BaseThingHandler {
     private final Runnable thingRefreshSender = () -> {
         try {
             triggerDeviceStatusRefresh();
-            if (!thing.getStatus().equals(ThingStatus.ONLINE)) {
-                updateStatus(ThingStatus.ONLINE);
-            }
+            updateStatus(ThingStatus.ONLINE);
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:104]: Btw, why was this changed? The same condition is in line 197, so the handler now has inconsistent logic for updating status to `ONLINE`.

File: bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeHandlerMock.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.openhab.core.thing.Thing;
+
+/**
+ * The {@link GoveeHandlerMock} is responsible for mocking {@link GoveeHandler}
+ * 
+ * @authorLeo Siepel - Initial contribution
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeHandlerMock.java:30]:
```suggestion
 * @author Leo Siepel - Initial contribution
```

File: bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandlerCallback;
+import org.openhab.core.types.State;
+
+import com.google.gson.Gson;
+
+/**
+ * @author Stefan Höhn - Initial contribution
```
Comment (jlaur) [inline bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:49]:
```suggestion
 * @author Leo Siepel - Initial contribution
```

(I assume)

**Review by jlaur — CHANGES_REQUESTED** [review]
Otherwise LGTM. @stefan-hoehn, would you like to review this fix?

File: bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java (hunk 12 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeSerializeGoveeHandlerTest.java:49]: yeah, copy paste.

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java (hunk 9 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/CommunicationManager.java:105]: Fixed

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java (hunk 10 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:104]: The framework already does this check in `ThingManagerImpl`, there is no need for this check. Bit minor though, without these checks in the bindings the code is cleaner. 
I also ran into a issue with the tests. I could fix it in the test, but prefer to remove this excess checking.

btw, i removed the other occurence too.

File: bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeHandlerMock.java (hunk 11 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeHandlerMock.java:30]: oopsie

Comment (lsiepel) [reply bundles/org.openhab.binding.govee/src/test/java/org/openhab/binding/govee/internal/GoveeHandlerMock.java]: oopsie

**Review by jlaur — APPROVED** [review]
LGTM

File: bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java [hunk 13]
```diff
@@ -134,7 +133,7 @@ public void initialize() {
         if (triggerStatusJob == null) {
             logger.debug("Starting refresh trigger job for thing {} ", thing.getLabel());
 
-            triggerStatusJob = scheduler.scheduleWithFixedDelay(thingRefreshSender, 100,
+            triggerStatusJob = executorService.scheduleWithFixedDelay(thingRefreshSender, 100,
```
Comment (stefan-hoehn) [inline bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:136]: Do you mind explaining why you replace the scheduler with the executerService?

Comment (stefan-hoehn) [reply bundles/org.openhab.binding.govee/src/main/java/org/openhab/binding/govee/internal/GoveeHandler.java:136]: Never mind, I went through the review now and saw the discussion. Sorry.

**Comment by stefan-hoehn** [comment]
@lsiepel Thanks for jumping in. You probably fixed it much better than I could have ever done it myself 🙏🏻

**Review by stefan-hoehn — APPROVED** [review]
